sbot/robot.py

```diff
@@ -1,189 +1,302 @@
+from .servo_board import ServoBoard
+from .utils import obtain_lock, singular
+
+logger = logging.getLogger(__name__)
+
+
+class Robot:
+    """
+    The main robot class that provides access to all the boards.
+
+    There can be only one instance of this class active in your operating
+    system at a time, creating a second instance will raise an OSError.
+
+    :param debug: Enable debug logging to the console, defaults to False
+    :param wait_start: Wait in the constructor until the start button is pressed,
+        defaults to True
+    :param trace_logging: Enable trace level logging to the console, defaults to False
+    :param manual_boards: A dictionary of board types to a list of serial port paths
+        to allow for connecting to boards that are not automatically detected, defaults to None
+    """
+    __slots__ = (
+        '_lock', '_metadata', '_power_board', '_motor_boards', '_servo_boards',
+        '_arduinos', '_cameras',
+    )
 
-from j5 import BaseRobot, BoardGroup
-from j5 import __version__ as j5_version
-from j5.backends import Backend, CommunicationError, Environment
-from j5.boards import Board
-from j5.boards.sb import SBArduinoBoard
-from j5.boards.sr.v4 import MotorBoard, PowerBoard, ServoBoard
-from j5.components.piezo import Note
-from j5_zoloto import ZolotoCameraBoard
+    def __init__(
+        self,
+        *,
+        debug: bool = False,
+        wait_start: bool = True,
+        trace_logging: bool = False,
+        manual_boards: dict[str, list[str]] | None = None,
+    ) -> None:
+        self._lock = obtain_lock()
+        self._metadata: Metadata | None = None
 
-from . import metadata
-from .env import HardwareEnvironment
-from .timeout import kill_after_delay
+        setup_logging(debug, trace_logging)
 
-__version__ = "0.10.1"
+        logger.info(f"SourceBots API v{__version__}")
 
-LOGGER = logging.getLogger(__name__)
+        if manual_boards:
+            self._init_power_board(manual_boards.get(PowerBoard.get_board_type(), []))
+            self._init_aux_boards(manual_boards)
+        else:
+            self._init_power_board()
+            self._init_aux_boards()
+        self._init_camera()
+        self._log_connected_boards()
 
-GAME_LENGTH = 120
+        if wait_start:
+            self.wait_start()
 
-BoardT = TypeVar("BoardT", bound=Board)
-BackendT = TypeVar("BackendT", bound=Backend)
+    def _init_power_board(self, manual_boards: list[str] | None = None) -> None:
+        """
+        Locate the PowerBoard and enable all the outputs to power the other boards.
 
+        :param manual_boards: Serial port paths to also check for power boards,
+            defaults to None
+        :raises RuntimeError: If exactly one PowerBoard is not found
+        """
+        power_boards = PowerBoard._get_supported_boards(manual_boards)
+        self._power_board = singular(power_boards)
+        self._power_board.outputs.power_on()
+        # TODO delay for boards to power up ???
```

Comment:
Do we want this delay?

Reply:
Need to try running sbot on a pi with all different boards powered from the power board to test this.

Reply:
I would recommend we merge this PR to make an initial robot image and then circle back later after testing the full image.